lib/cpp/prox/prox_slope.cpp

```diff
@@ -2,34 +2,36 @@
 
 #include "tick/prox/prox_slope.h"
 
-ProxSlope::ProxSlope(double strength,
-                     double false_discovery_rate,
-                     bool positive)
-  : ProxSortedL1(strength, WeightsType::bh, positive) {
+template <class T>
+TProxSlope<T>::TProxSlope(T strength, T false_discovery_rate, bool positive)
+    : TProxSortedL1<T>(strength, WeightsType::bh, positive) {
   this->false_discovery_rate = false_discovery_rate;
 }
 
-ProxSlope::ProxSlope(double strength,
-                     double false_discovery_rate,
-                     ulong start,
-                     ulong end,
-                     bool positive)
-  : ProxSortedL1(strength, WeightsType::bh, start, end, positive) {
+template <class T>
+TProxSlope<T>::TProxSlope(T strength, T false_discovery_rate, ulong start,
+                          ulong end, bool positive)
+    : TProxSortedL1<T>(strength, WeightsType::bh, start, end, positive) {
   this->false_discovery_rate = false_discovery_rate;
 }
 
-const std::string ProxSlope::get_class_name() const {
-  return "ProxSlope";
+template <class T>
+std::string TProxSlope<T>::get_class_name() const {
+  return "TProxSlope";
```

Comment:
Ah ah here get_class_name returns a string :) Can you do the same as for models ? Namely remove all of them and get the name in the main parent class using idtype ?

Reply:
can we address all "get_class_name" with a single PR?